hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/management/JpaDistributionSetManagement.java

```diff
@@ -494,51 +549,62 @@ private Specification<JpaDistributionSetMetadata> byDsIdSpec(final long dsId) {
+        assertDistributionSetExists(id);
 
-        return distributionSetMetadataRepository.countByDistributionSetId(setId);
+        return distributionSetMetadataRepository.countByDistributionSetId(id);
     }
 
     @Override
     public Page<DistributionSetMetadata> findMetaDataByDistributionSetIdAndRsql(final Pageable pageable,
-            final long distributionSetId, final String rsqlParam) {
-        throwExceptionIfDistributionSetDoesNotExist(distributionSetId);
+                                                                                final long id, final String rsqlParam) {
+        assertDistributionSetExists(id);
 
         final List<Specification<JpaDistributionSetMetadata>> specList = Arrays
                 .asList(RSQLUtility.buildRsqlSpecification(rsqlParam, DistributionSetMetadataFields.class,
-                        virtualPropertyReplacer, database), byDsIdSpec(distributionSetId));
+                        virtualPropertyReplacer, database), byDsIdSpec(id));
 
         return JpaManagementHelper.findAllWithCountBySpec(distributionSetMetadataRepository, pageable, specList);
     }
 
     @Override
-    public Optional<DistributionSetMetadata> getMetaDataByDistributionSetId(final long setId, final String key) {
-        throwExceptionIfDistributionSetDoesNotExist(setId);
+    public Optional<DistributionSetMetadata> getMetaDataByDistributionSetId(final long id, final String key) {
+        assertDistributionSetExists(id);
 
-        return distributionSetMetadataRepository.findById(new DsMetadataCompositeKey(setId, key))
+        return distributionSetMetadataRepository
+                .findById(new DsMetadataCompositeKey(id, key))
                 .map(DistributionSetMetadata.class::cast);
     }
 
     @Override
     public Optional<DistributionSet> getByAction(final long actionId) {
-        if (!actionRepository.existsById(actionId)) {
-            throw new EntityNotFoundException(Action.class, actionId);
-        }
-
-        return Optional.ofNullable(distributionSetRepository.findByActionId(actionId));
+        return actionRepository
+                .findById(actionId)
+                .map(action -> {
+                    if (!targetRepository.exists(TargetSpecifications.hasId(action.getTarget().getId()))) {
+                        throw new InsufficientPermissionException("Target not accessible (or not found)!");
+                    }
+                    return distributionSetRepository
+                            .findOne(DistributionSetSpecification.byActionId(actionId))
+                            .orElseThrow(() ->
+                                    new InsufficientPermissionException("DistributionSet not accessible (or not found)!"));
+                })
+                .map(DistributionSet.class::cast)
+                .or(() -> {
+                    throw new EntityNotFoundException(Action.class, actionId);
+                });
```

Comment:
In generell I would not throw an InsufficientPermissionException in case of missing read rights. Instead, I would filter the stream (is target found) and provide an empty Optional in case the associate target is not accessible.

Reply:
Example:

```suggestion
        return actionRepository.findById(actionId)
                .filter(action -> targetRepository.exists(TargetSpecifications.hasId(action.getTarget().getId())))
                .map(action -> distributionSetRepository.findOne(DistributionSetSpecification.byActionId(actionId))
                        .orElseThrow(() -> new InsufficientPermissionException(
                                "DistributionSet not accessible (or not found)!")))
                .map(DistributionSet.class::cast).or(() -> {
                    throw new EntityNotFoundException(Action.class, actionId);
                });
```

Reply:
First of all, I think that returning optional methods shall not throw not found. So, if not found shall return empty,
However, this would change the behaviour. So ..
Regarding the insufficient permissions - currently, since there is no action access control I assume that the access control for actions is provided by targets. Thus, if no access to target I assume no access to action. That's why I think here shall be thrown insufficient permissions instead of not found